text/0059-functions.md

```diff
@@ -0,0 +1,282 @@
+All functions in a policyset are in scope for all policies. (ie: function declarations do not need to lexically proceed use)
+Cedar policies and function bodies are lexically scoped.
+`principal`/`action`/`resource`/`context` are considered to be more tightly bound then function names.
+This means that while you could name a function `principal`, you could never call it. (This should probably be a validator warning)
+Function name conflicts at the top level are a parse error.
+Function names may shadow extension functions (Results in a warning).
+
+### Formal Semantics
+This RFC adds one new evaluation rule to Cedar:
+If $f$ is the name of a declared function, $def(f)$ is the body of the definition, and $p_1, ..., p_n$ is the list of parameters in the definition:
+$$ f(v_1, ..., v_n) \rightarrow def(f) / [p_1 \leftarrow v_1, ..., p_n \leftarrow v_n] $$
+Where $/$ is standard substitution.
+
+### Formal Grammar
+The grammar of Cedar expressions is unchanged, as we re-use the existing call form.
+```
+function ::= 'function' Path '(' Params? ')' '{' Expr '}' ';'
+Params ::= Ident (',' Ident)? ','?
+```
+(Allows trailing commas in parameter lists)
+### Typechecking
+The validator can typecheck functions at use-sites via inlining.
+
+## Drawbacks
+
+### Redundancy
+Functions in this RFC can only accomplish things that can already accomplished with Cedar expressions. 
+This means we are not expanding the expressive power of Cedar in any way. 
+We are also adding more than one way to accomplish the same task.
+Brining back our SemVer example:
+```
+// Permit access when the api is greater than 2.1
+permit (principal, action == Action::"someAction", resource)
+when {
+    resource.apiVersion.major > 2 ||
+    (resource.apiVersion.major == 2 && resource.apiVersion.minor >= 1)
+};
+```
+This policy does accomplish the user's goal of encoding the SemVer relationship.
+The problem here is readability. A reader of this policy has to identify that this is implementing the standard semantic version comparison operator, and not some bespoke versioning scheme. It’s also hurts maintainability, especially if you have multiple policies that reason about SemVers, you have to repeat the logic in each policy. This violates some basic software engineering tenets (Don’t Repeat Yourself), allowing bugs to sneak in via typos or botched copy/paste jobs. 
+
+### No Custom Parsing
+Using extension functions lets you implement a custom parser for your datatype. 
+This is useful even if your data type can be expressed as pure Cedar.
+For example, both the (existing) decimal expression and (proposed) semver extension make use of this:
+```
+decimal("1.1")
+semver("2.3.5")
+```
+vs
+```
+decimal(1,1)
+semver(2,3,5)
+```
+
+### Readability/Policies are no longer standalone
+A policy can no longer be read by itself, it has to be read in the context of all function definitions it uses.
+Policies that use an inordinate number of functions may be very opaque to read.
+
```

Comment:
Another downside of this proposal is that it's unclear (a) where the functions are defined and how they are managed, (b) how they are passed to Cedar at runtime, and (c) which functions are available when to what policies. 

In a simple world where all Cedar policies are stored in a single file, we can put the functions in the same file.

But what about a hosted service, where policies are stored in a DB, and the service manages policies for many different applications and clients? 

What functions are available to what clients?  Where are they stored?  How do clients discover functions?  Does the service have to effectively do runtime linking, or are functions inlined at typechecking / creation type?  Does the service have to gather the set of functions to pass in to Cedar for evaluation?

Reply:
My perspective is that these are good questions but outside the scope of this RFC. At least for now, I propose we just support functions as being definable in the same file as policies. 

For context, I note we have similar questions around how to store and manage the "Cedar standard library" from [RFC 58](https://github.com/cedar-policy/rfcs/blob/cdisselkoen/standard-library/text/0058-standard-library.md). We also don't solve the problem of bringing together policies authored separately, e.g., by an app owner and its users. We can leave solutions of these questions for later, and perhaps even outside the language.

Reply:
> execution of even a single policy if the problem is not solved.
There are solutions to this. In the absolute absense of being able to load a function context they can be inlined. (Esp if we take the CBN semantics below, which I agree with)

Reply:
Decimal is already solved by the `Decimal` extension type.  

As for Datetime, I'm having a hard time thinking of non-trivial Datetime functions that can be easily expressed in Cedar and for which this would be a good solution.  (Though that could be a lack of imagination on my part :)

For example, consider a function that takes as input a timestamp and calculates which day of the week this corresponds to.  This is a hairy complicated calculation and it would be very painful to write in Cedar (if it's even possible---I genuinely don't know if this algorithm can be expressed with loops that can be feasibly unrolled).

Also, inlining a function like this everywhere would not scale for analysis.  In fact, even if we don't inline them explicitly, that is effectively what the analysis will have to end up doing anyway.  (The encoding of the function's body will be inlined at each callsite---potentially with specific sequencing of argument evaluation if we want CBV.). So, while we can analyze Cedar functions in theory, our practical ability to do so will very much hinge on these functions between small and simple.

My other concern is that this proposal is a non-trivial breaking change to the language semantics and APIs.   We shouldn't underestimate just how pervasive the changes must be to implement this as proposed.

For example, to evaluate policies, the user now has to pass in an environment that binds function names to their implementations (at least as described in this RFC).  So, in addition to policy slicing and entity slicing, users may have to potentially do function slicing (i.e., figure out which functions to pass in if they have thousands of them scattered in various places).

Finally, this RFC is going to make our formalization and proofs significantly more complicated---especially deep semantic proofs that have to reason about termination of evaluation, typechecking, and symbolic compilation in the presence of functions.  We should not underestimate the proof engineering difficulty that this brings. 

In my view, we are yet to see a sufficiently compelling use case to justify adding this level of complexity to the language, formalization, analysis, and implementation.

Reply:
> So, while we can analyze Cedar functions in theory, our practical ability to do so will very much hinge on these functions between small and simple.
Isn't this also the case for cedar policies as they exist now? If people aren't using functions now and just writing everything out manually it's the same thing. The only argument is that writing functions makes it easier (and therefor more attractive) to write bigger policies, and given the fact that functions can't call other functions that seems weaker.

> Finally, this RFC is going to make our formalization and proofs significantly more complicated---especially deep semantic proofs that have to reason about termination of evaluation, typechecking, and symbolic compilation in the presence of functions. We should not underestimate the proof engineering difficulty that this brings.

Is this true if you inline things? I don't see why it would be.

Reply:
> Decimal is already solved by the Decimal extension type.
Yes but we've talked before about how we don't love that it's an extension since it's expressible in pure Cedar. my point was that _would_ have solved this need without needing an extension.

Reply:
My understanding of the proposal is that we want to support actual functions (lambdas) here, which involves passing in an environment that binds function names to bodies and evaluating all this on the fly---as well as adjusting the semantics, formalization, etc., accordingly.  

If we aren't proposing functions, and we are proposing macros, then formalization etc don't have to change.

But that's a different proposal and we have to worry about different things (hygiene etc).

Reply:
Yup, these are basically extremely restricted macros. No need for bindings on the fly. Hygiene is trivial since macros can't invoke other macros, and policies only bind `PARC`

Reply:
I think it's worth revising the proposal to make this clear :). 

I would also clarify that the formal semantics, AST, etc. don't actually change in an any way with this proposal.  If these are macros, then they desugar to core Cedar via a pre-processing pass, and we don't need a new evaluation rule for them:

https://github.com/cedar-policy/rfcs/blob/aaronjeline/functions/text/0061-functions.md#formal-semantics

Instead, we provide expansion / desugaring rules that are separate from the formal semantics of the language.

Reply:
Good callout

Reply:
(And we can separately prove facts about desugaring.)

Reply:
These macros can't create policies, whereas templates can. Coming up with a unified macro for both levels of the language sound annoying, but maybe im just not seeing it.

Reply:
Suppose we adjust this RFC's syntax to be
```
def semver(?major, ?minor, ?long) {
  { major : ?major, minor : ?minor, patch : ?patch }
};
```
That is
- We change `function` to `def`
- We require that macro parameter variables are prefixed with a `?`

I think these changes convey the nature of what we are providing as a macro: we `def` a macro, and `?`-prefixed parameters could be expressions, not just values. The change to variable names also resolves the thorny issue around name shadowing -- we don't have to worry about a parameter shadowing the `principal` variable for example, which is "global." With this change, we can allow macros to freely reference it.

I also think this syntax opens the door to more generalized templates. The `?var` parameters are the same in both. Future syntax for generalized templates could be like the following (which evokes one of the alternatives of [RFC 3](https://github.com/cedar-policy/rfcs/blob/placeholders-in-conditions/text/0003-placeholders-in-conditions.md), which was rejected until more was known):
```
def template cross_relation(?user, ?resourcebound)
  permit(principal, action, resource)
  when {
       ?resourcebound in resource
    && principal in ?user.delegates
  };
```
I'm not saying we include templates in this RFC; I'm saying that taking this RFC could gel nicely with a future generalized templates RFC.

Reply:
And while I'm at it, let me suggest a different option for the syntax: Drop the `{` `}` around a macro definition. So we would have:
```
def semver(?major, ?minor, ?long)
  { major : ?major, minor : ?minor, patch : ?patch }
;
```
rather than
```
def semver(?major, ?minor, ?long) {
  { major : ?major, minor : ?minor, patch : ?patch }
};
```
Arguing in favor
1. Cedar expressions don't have curly braces around them, and dropping curly braces around the definition makes returning a record less ugly/error prone, e.g., `def semver(...) { ... };` instead of `define semver(...) { { ... } };`
2. Short macros will look better without the braces, e.g., `def ismanager(?e1,?e2) ?e1.manager == ?e2` vs. `def ismanager(?e1,?e2) { ?e1.manager == ?e2 }`
3. The use of `;` as a terminator looks more natural after an expression than it does after a closing curly brace of a top-level definition, i.e., `};` -- no other language does this, that I know of.

Arguing against:
1. The curly braces, while confusable with record literal syntax, may make the macro body easier to visually isolate. I can't think of better delimiters, but maybe you could use a separator instead. For example, if you use `:=` we'd have `def ismanager(?e1,?e2) := ?e1.manager == ?e2;` but I worry this is not easier to read and is no less confusing

Another orthogonal syntax idea is to change `def` to `define`, to evoke C-style macros, which developers will be familiar with. I don't like this as much though because (a) C-style macros need to be all on one line, which Cedar macros don't; (b) C-style macros don't have a special `?var` syntax for parameters; and (c) they don't enforce hygiene. Also, I like that `def` is sort of like `define`, but also is used to define functions in Ruby and Python. Our proposed macros are basically call-by-name functions, so this is not a terrible analogy to make.

Considering other language syntaxes for inspiration
- [Scala has macros](https://docs.scala-lang.org/overviews/macros/overview.html), and they use `def` as the keyword to introduce them, as with normal functions, but then add keyword `macro` prior to the body. So something like `def ismanager(?e1,?e2) macro ?e1.manager == ?e2;`. Looks a little odd but it could grow on me.
- [Scheme/LISP macros](https://en.wikibooks.org/wiki/Scheme_Programming/Macros) are the benchmark, but the S-expression syntax doesn't carry over. They use keyword `defmacro` or `define-syntax`.
- [Julia has macros](https://docs.julialang.org/en/v1/manual/metaprogramming/#man-macros) but it's not a good stylistic fit (and few people will know it)
- [Rust has macros](https://doc.rust-lang.org/book/ch19-06-macros.html) but again the syntax is not a great stylistic fit (e.g., use of `$` and `!` in the def, and `#` in annotations), and they are much more powerful

Reply:
Updated the RFC. Removed the section on var-vs-function namespaces since it's now moot